### ISSUE TITLE
[in_app_purchase] Cannot get trial period of subscription

### ISSUE THREAD

**Issue by RomanSoviak** [flutter/flutter#143514]
### Steps to reproduce

I got subscriptions info details after request `queryProductDetails` method, but I cannot find the trial period for the subscriptions. My subscriptions have trials it, but I cannot fetch that trials.

### Expected results

I want to get a trial period for the subscription

### Actual results

Object `ProductDetailsResponse` doesn't contain trial period

### Code sample

```
    final productIDs = [
      PREMIUM_SUBSCRIPTION_MONTHLY,
      PREMIUM_SUBSCRIPTION_YEARLY,
      PREMIUM_LIFETIME
    ];

    final _inAppPurchase = InAppPurchase.instance;
   _inAppPurchase.queryProductDetails(productIDs.toSet()).then((value) {
     print(value);
   });
```

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]
<img width="449" alt="Screenshot 2024-02-15 at 11 29 28" src="https://github.com/flutter/flutter/assets/26696190/09685a2f-5526-41dc-a6c1-f1f7a23986ee">

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by dam-ease** [comment]
Hello @RomanSoviak.
Thanks for the report!
This seems similar/related to https://github.com/flutter/flutter/issues/43720. 
Can you please check this out?

**Comment by snoopdoggy322** [comment]
I encountered the same problem, in addition, the trial subscription to Google is like another subscription tariff that must be purchased as a full-fledged product, instead of, as if interacting with Apple, buying the same product that is covered by the trial version.
Here is a code example to easily get the trial validity period for apple:

```
if (product is AppStoreProductDetails) {
      if (product.skProduct.introductoryPrice != null) {
        var introductoryPrice = product.skProduct.introductoryPrice!;
        return introductoryPrice.subscriptionPeriod.numberOfUnits;
      }
    }
```

I still haven’t figured out how to get the validity period of a trial subscription for a product on Android.

**Comment by RomanSoviak** [comment]
@snoopdoggy322 thanks for the reply! I saved the free trial period for the Android on our server. And make the call to the server, but it would be good to fetch it from the purchase lib 🙂

**Comment by snoopdoggy322** [comment]
@RomanSoviak As one of the options, but it seems that it will not solve the problem with purchasing such a subscription.
Are you able to buy a subscription with a trial period? So that it would be displayed in the purchase window from Google Play itself?

**Comment by RomanSoviak** [comment]
@snoopdoggy322 Sorry for the late response I switched to another Flutter purchase library, here I can do it

**Comment by mikron123** [comment]
> @RomanSoviak As one of the options, but it seems that it will not solve the problem with purchasing such a subscription. Are you able to buy a subscription with a trial period? So that it would be displayed in the purchase window from Google Play itself?

Having the same problem. Upgraded the package version and now clients are not getting the trial period offer, and are charged immediately!

There's no documentation about how to include the offer, is it a bug?

**Comment by Zachinio** [comment]
This is issue is now a blocker for my version as well, I offer a free trial for my users without any backend.
Now Google rejects my app updates claiming that my app offers a free trial without actually offering it.

**Comment by mikron123** [comment]
> This is issue is now a blocker for my version as well, I offer a free trial for my users without any backend. Now Google rejects my app updates claiming that my app offers a free trial without actually offering it.

Same here

**Comment by jiang1654** [comment]
Since version 3.1.5, Android devices have not been able to receive free trials

**Comment by CLIDOapps** [comment]
Hi, this issue is a blocker for my Android app too.

Please fix this plugin.

The code in the migration guide is not working:

`/// Handles the introductory price period of a subscription.
void handleIntroductoryPricePeriod(ProductDetails productDetails) {
  if (productDetails is GooglePlayProductDetails) {
    final ProductDetailsWrapper product = productDetails.productDetails;
    if (product.productType == ProductType.subs) {
      // Unwrapping is safe because the product is a subscription.
      final SubscriptionOfferDetailsWrapper offer =
          product.subscriptionOfferDetails![productDetails.subscriptionIndex!];
      final List<PricingPhaseWrapper> pricingPhases = offer.pricingPhases;
      if (pricingPhases.length >= 2 &&
          pricingPhases.first.priceAmountMicros <
              pricingPhases[1].priceAmountMicros) {
        // Introductory pricing period logic.
      }
    }
  }
}`

pricingPhases.length is on Android devices always "1".

**Comment by Meta-Learn-ApS** [comment]
Does anyone have a fix for this? I get an error when trying to downgrade from 3.2.0 to 3.1.5, or do we know if a fix is on the way?

**Comment by andynvt** [comment]
same here

**Comment by joselicht90** [comment]
is there any workaround to this? that doesn't include a backend server.

**Comment by DusBekas** [comment]
Is there any solution for this issue?

So far I can get the offers but I don't know how to implement the free trial logic as mentioned in this [example](https://github.com/flutter/packages/blob/main/packages/in_app_purchase/in_app_purchase_android/example/lib/migration_guide_examples.dart)

Here is my code:
`for (final SubscriptionOfferDetailsWrapper offer in product.subscriptionOfferDetails!) {
        print('Offer Details:');
        print('  - Offer ID: ${offer.offerId}');
        print('  - Pricing Phases:');

        for (final PricingPhaseWrapper pricingPhase in offer.pricingPhases) {
          print('    - Price: ${pricingPhase.priceAmountMicros}');
          print('    - Duration: ${pricingPhase.billingPeriod}');
        }

        if (offer.offerId == 'freetrial') {
          // implement the logic
        }
      }`

With this code I get:
> I/flutter (11052): Offer Details:
I/flutter (11052):   - Offer ID: freetrial
I/flutter (11052):   - Pricing Phases:
I/flutter (11052):     - Price: 0
I/flutter (11052):     - Duration: P1W
I/flutter (11052):     - Price: 120000000000
I/flutter (11052):     - Duration: P1M
I/flutter (11052): Offer Details:
I/flutter (11052):   - Offer ID: null
I/flutter (11052):   - Pricing Phases:
I/flutter (11052):     - Price: 120000000000
I/flutter (11052):     - Duration: P1M
I/flutter (11052): Offer Details:
I/flutter (11052):   - Offer ID: freetrial
I/flutter (11052):   - Pricing Phases:
I/flutter (11052):     - Price: 0
I/flutter (11052):     - Duration: P1W
I/flutter (11052):     - Price: 120000000000
I/flutter (11052):     - Duration: P1M
I/flutter (11052): Offer Details:
I/flutter (11052):   - Offer ID: null
I/flutter (11052):   - Pricing Phases:
I/flutter (11052):     - Price: 120000000000
I/flutter (11052):     - Duration: P1M


Basically it is a monthly subscription with 7 days free trial. I don't get how we could implement these trial/offers to actually buy the subscription with a proper trial. 

Also as far as my testing went, on iOS the free trial just showing without any further intervention.

**Comment by Sotatek-PhiPham** [comment]
> Since version 3.1.5, Android devices have been unable to receive free trials

After downgrading to 3.1.4, the payment popup correctly displays "Trial" as expected on android. But we can use version 3.1.4 because it uses old billing version

**Comment by Sotatek-PhiPham** [comment]
@stuartmorgan  Please take a look at this—my client is losing users because of this issue. Thanks

**Comment by mikron123** [comment]
> > Since version 3.1.5, Android devices have been unable to receive free trials
> 
> After downgrading to 3.1.4, the payment popup correctly displays "Trial" as expected on android. But we can use version 3.1.4 because it uses old billing version

The solution is to scan all products which return from the IAP. You will get separate products for the trial ones, with same ID but  price 0. You will need to call purchase() with them in order to initiate the trial.

**Comment by parthbhensdadiya7** [comment]
Hi @darshankawar, any update on this?

**Comment by Hari-07** [comment]
On ios, you can do 
```dart
final details = productDetailResponse.productDetails.first;
    if (details is AppStoreProduct2Details) {
      print(details.sk2Product.subscription?.subscriptionPeriod);
    }
```
